onnxruntime/core/providers/nuphar/common/analysis/analysis.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/codegen/common/common.h"
+#include "core/common/common.h"
+#include "core/graph/graph_viewer.h"
+#include "core/providers/nuphar/common/nuphar_subgraph.h"
+
+namespace onnxruntime {
+namespace nuphar {
+
+// abstract class for Analysis
+template <typename INPUT_TYPE>
+class AnalysisBase {
+ public:
+  AnalysisBase() {}
+
+  AnalysisBase(const std::string& name)
+      : name_(name) {}
+
+  virtual ~AnalysisBase() = default;
+
+  virtual void Evaluate(INPUT_TYPE) = 0;
+
+  const std::string& Name() const {
+    return name_;
+  }
+
+ protected:
+  std::string name_{"Unknown"};
+
```

Comment:
If a derived class need to update this, provide a protected method to do so instead of giving them direct access.

Reply:
This should only be updated by derived class in ctor, changing it to const.

---
In reply to: [313192277](https://github.com/microsoft/onnxruntime/pull/881#discussion_r313192277) [](ancestors = 313192277)